src/workerd/io/actor-sqlite.c++

```diff
@@ -383,11 +511,37 @@ void ActorSqlite::shutdown(kj::Maybe<const kj::Exception&> maybeException) {
 }
 
 kj::Maybe<kj::Own<void>> ActorSqlite::armAlarmHandler(kj::Date scheduledTime, bool noCache) {
-  return hooks.armAlarmHandler(scheduledTime, noCache);
+  KJ_ASSERT(!haveDeferredDelete);
+  KJ_ASSERT(!inAlarmHandler);
+
+  auto localAlarmState = metadata.getAlarm();
+  if (localAlarmState != scheduledTime) {
```

Comment:
If we receive a premature alarm (`scheduledTime < localAlarmState`) then we need to update the alarm manager to the correct time, right?

Reply:
To be clear, the scenario would be:
* Application updates alarm from X to Y, where Y is later than X.
* The database commit goes through and is persisted.
* The alarm update (which happens after the commit, since it's moving later) fails.
* The machine dies.
* Later, the alarm fires at time X. At this point, no writes/commits/etc. are in progress or anything.
* We need to respond by updating the alarm time to Y (and skip running the alarm for now).

Reply:
So, I'd been thinking through how to implement this, and I think it might require a slight change in API to do it correctly...  The issue is that `armAlarmHandler()` can only return either `kj::none`, which cancels the alarm without retry, or the `kj::Own<void>` deleter value.  If we issue the alarm rescheduling request asynchronously and return `kj::none`, I think it potentially leaves a window where the AlarmManager will no longer try the old alarm time but the new alarm time has not yet been confirmed.

I'm thinking of adding a `kj::Promise<void>` to the set of types that `armAlarmManager()` can return, such that the caller can wait for the rescheduling to be successful before returning a canceled result.

Reply:
Implemented in 3a10dcfd .